### DISCUSSION TITLE
Advanced Query Param Parsing Using Depends

### DISCUSSION THREAD

**Discussion by jordantshaw** [fastapi/fastapi#3180] [Questions] [answered]
Hi all,
I am looking to implement some more advanced query parameter options where users can specify if they want the parameter to be lt, lte, etc.... I am planning on implementing this the same way Stripe API does it, using dot notation (see example below).
https://stripe.com/docs/api/products/list#list_products-created-gte
Example: created, created.gt, created.gte, etc...

Currently I am trying to implement this using the Depends class to dynamically create the query params.  I was hoping to do something like the following.  But obviously I can't use the self.param_name in the alias field because self is not defined yet. So I would be forced to create a new function or class for every parameter I want to do this with.  

```
class AdvancedQueryParser:
    def __init__(self, param_name: str):
        self.param_name = param_name

    def __call__(
            self,
            eq: date = Query(None, alias=f'{self.param_name}'),
            gte: date = Query(None, alias=f'{self.param_name}.gte'),
            lte: date = Query(None, alias=f'{self.param_name}.lte'),
    ):

        return {
            'eq': eq,
            'gte': gte,
            'lte': lte,
        }


@router.get('/items')
def get_items(
        created: dict = Depends(AdvancedQueryParser('created')),
        token_payload: schemas.TokenPayload = Depends(security.verify_access_token)
) -> Any:
    pass
```

Does anyone know of a better way to implement this? I would really like to avoid having to create multiple functions/classes for every parameter that needs to use these advanced options, or worse, be forced to just list all the possible query parameters in the function definition itself.

**Comment by mxp-xc** [comment, marked as answer]
You can use function instead of class
```
import functools
from typing import Any

from fastapi import Query, FastAPI, Depends

app = FastAPI()


@functools.lru_cache
def get_advanced_query_parser(param_name: str, data_type):
    def parser(
            eq: data_type = Query(None, alias=param_name),
            gte: data_type = Query(None, alias=f"{param_name}.gte"),
            lte: data_type = Query(None, alias=f"{param_name}.lte"),
    ):
        return {
            'eq': eq,
            'gte': gte,
            'lte': lte,
        }

    return parser


@app.get('/items')
def get_items(
        created: dict = Depends(get_advanced_query_parser('created', int)),
        t1: dict = Depends(get_advanced_query_parser('t1', str)),
) -> Any:
    return {"created": created, "t1": t1}


if __name__ == '__main__':
    import uvicorn

    uvicorn.run("test13:app", use_colors=True)

```

**Reply by jordantshaw** [reply]
This is a great solution, thank you!
